File: 2021/2021W39/preppindataW39.R
library(readr)
library(dplyr)
library(tidyr)
library(stringr)

final <- read_csv("C:/Data/PreppinData/Bike Painting Process - Painting Process.csv") %>%
  mutate('Datetime' = as.POSIXct(paste(`Date`, `Time`),format='%d/%m/%Y %H:%M:%S'),
         'Bike Type' = if_else(`Data Parameter`=='Bike Type', `Data Value`, NA_character_),
         'Batch Status' = if_else(`Data Parameter`=='Batch Status', `Data Value`, NA_character_),
         'Name of Process Stage' = if_else(`Data Parameter`=='Name of Process Stage', `Data Value`, NA_character_)) %>%
  group_by(`Batch No.`) %>%
  arrange(`Datetime`) %>%
  fill(`Bike Type`, `Batch Status`, `Name of Process Stage`) %>%
  filter((`Data Type`=='Process Data') & (`Data Parameter`!='Name of Process Stage')) %>%
  mutate('Actual' = if_else(str_detect(`Data Parameter`, '^(Actual)'), as.numeric(`Data Value`), NA_real_),
         'Target' = if_else(str_detect(`Data Parameter`, '^(Target)'), as.numeric(`Data Value`), NA_real_),
         'Data Parameter' = str_remove(`Data Parameter`,'^(\\w+\\s)')) %>%
  select(c('Batch No.', 'Bike Type', 'Batch Status', 'Name of Process Stage',
           'Data Parameter', 'Actual', 'Target', 'Datetime'))

View(final)

File: 2021/2021W39/PreppindataW39.py
import pandas as pd
import re

process = pd.read_csv("C:\\Data\\PreppinData\\Bike Painting Process - Painting Process.csv")
process['Datetime'] = process.apply(lambda x: pd.to_datetime(f"{x['Date']} {x['Time']}"), axis=1)
process = process.groupby(['Batch No.']).apply(pd.DataFrame.sort_values, 'Datetime', ascending=True).reset_index(drop=True)
 
for attr in ['Bike Type', 'Batch Status', 'Name of Process Stage']:
    process[attr] = process.apply(lambda x: x['Data Value'] if x['Data Parameter']==attr else None, axis=1)
    process[attr] = process.groupby(['Batch No.'])[attr].ffill() 
    
process = process[(process['Data Type']=='Process Data') & (process['Data Parameter']!='Name of Process Stage')]
for vtype in ['Actual', 'Target']:
    process[vtype] = process.apply(lambda x: float(x['Data Value']) if re.match(f'^({vtype})',x['Data Parameter']) else None, axis=1)
process['Data Parameter'] = process['Data Parameter'].apply(lambda x: re.sub('^(\w+\s)','',x))
process = process[['Batch No.', 'Bike Type', 'Batch Status', 'Name of Process Stage',
                   'Data Parameter', 'Actual', 'Target', 'Datetime']]
   

